File: dev-server.js
const mockMetrics = `# HELP tor_snowflake_proxy_connections_total The total number of successful connections handled by the snowflake proxy
# TYPE tor_snowflake_proxy_connections_total counter
tor_snowflake_proxy_connections_total{country=""} 3
tor_snowflake_proxy_connections_total{country="CI"} 1
tor_snowflake_proxy_connections_total{country="ES"} 1
tor_snowflake_proxy_connections_total{country="FR"} 1
tor_snowflake_proxy_connections_total{country="GB"} 2
tor_snowflake_proxy_connections_total{country="IR"} 9
tor_snowflake_proxy_connections_total{country="IT"} 1
tor_snowflake_proxy_connections_total{country="JP"} 1
tor_snowflake_proxy_connections_total{country="KR"} 1
tor_snowflake_proxy_connections_total{country="MA"} 1
tor_snowflake_proxy_connections_total{country="MU"} 2
tor_snowflake_proxy_connections_total{country="NG"} 2
tor_snowflake_proxy_connections_total{country="PG"} 1
tor_snowflake_proxy_connections_total{country="PT"} 1
tor_snowflake_proxy_connections_total{country="RU"} 7
tor_snowflake_proxy_connections_total{country="SG"} 1
tor_snowflake_proxy_connections_total{country="TN"} 3
tor_snowflake_proxy_connections_total{country="US"} 41
# HELP tor_snowflake_proxy_connection_timeouts_total The total number of client connection attempts that failed after successful rendezvous
# TYPE tor_snowflake_proxy_connection_timeouts_total counter
tor_snowflake_proxy_connection_timeouts_total 91
# HELP tor_snowflake_proxy_traffic_inbound_bytes_total The total in bound traffic by the snowflake proxy (KB)
# TYPE tor_snowflake_proxy_traffic_inbound_bytes_total counter
tor_snowflake_proxy_traffic_inbound_bytes_total 490539
# HELP tor_snowflake_proxy_traffic_outbound_bytes_total The total out bound traffic by the snowflake proxy (KB)
# TYPE tor_snowflake_proxy_traffic_outbound_bytes_total counter
tor_snowflake_proxy_traffic_outbound_bytes_total 110937
# HELP process_network_receive_bytes_total Number of bytes received by the process over the network
# TYPE process_network_receive_bytes_total counter
process_network_receive_bytes_total 1.78163309e+09
# HELP process_network_transmit_bytes_total Number of bytes sent by the process over the network
# TYPE process_network_transmit_bytes_total counter
process_network_transmit_bytes_total 1.780049424e+09
# HELP process_resident_memory_bytes Resident memory size in bytes
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 9.8537472e+07
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.7695821806e+09
# HELP go_goroutines Number of goroutines that currently exist
# TYPE go_goroutines gauge
go_goroutines 571
# HELP process_open_fds Number of open file descriptors
# TYPE process_open_fds gauge
process_open_fds 65
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 1395.02
`;

const server = Bun.serve({
	port: 3000,
	async fetch(req) {
		const url = new URL(req.url);

		if (url.pathname === "/internal/metrics") {
			return new Response(mockMetrics, {
				headers: { "Content-Type": "text/plain" },
			});
		}

		if (url.pathname === "/internal/logs") {
			const mockLogs = `2026/01/29 06:36:23 In the last 1h0m0s, there were 28 completed successful connections. Traffic Relayed ↓ 198456 KB (55.13 KB/s), ↑ 38921 KB (10.81 KB/s).
2026/01/29 05:36:23 In the last 1h0m0s, there were 35 completed successful connections. Traffic Relayed ↓ 251032 KB (69.73 KB/s), ↑ 47852 KB (13.29 KB/s).
2026/01/29 04:36:23 In the last 1h0m0s, there were 42 completed successful connections. Traffic Relayed ↓ 312847 KB (86.90 KB/s), ↑ 59234 KB (16.45 KB/s).
2026/01/29 03:36:23 In the last 1h0m0s, there were 31 completed successful connections. Traffic Relayed ↓ 221543 KB (61.54 KB/s), ↑ 41876 KB (11.63 KB/s).
2026/01/29 02:36:23 In the last 1h0m0s, there were 19 completed successful connections. Traffic Relayed ↓ 134219 KB (37.28 KB/s), ↑ 25183 KB (6.99 KB/s).
2026/01/29 01:36:23 In the last 1h0m0s, there were 47 completed successful connections. Traffic Relayed ↓ 341876 KB (94.97 KB/s), ↑ 64392 KB (17.89 KB/s).
2026/01/29 00:36:23 In the last 1h0m0s, there were 38 completed successful connections. Traffic Relayed ↓ 267543 KB (74.32 KB/s), ↑ 50124 KB (13.92 KB/s).
2026/01/28 23:36:23 In the last 1h0m0s, there were 52 completed successful connections. Traffic Relayed ↓ 389234 KB (108.12 KB/s), ↑ 72841 KB (20.23 KB/s).
2026/01/28 22:36:23 In the last 1h0m0s, there were 44 completed successful connections. Traffic Relayed ↓ 298712 KB (82.98 KB/s), ↑ 56234 KB (15.62 KB/s).
2026/01/28 21:36:23 In the last 1h0m0s, there were 29 completed successful connections. Traffic Relayed ↓ 203487 KB (56.52 KB/s), ↑ 38291 KB (10.64 KB/s).
2026/01/28 20:36:23 In the last 1h0m0s, there were 36 completed successful connections. Traffic Relayed ↓ 254098 KB (70.58 KB/s), ↑ 47659 KB (13.24 KB/s).
2026/01/28 19:36:23 In the last 1h0m0s, there were 41 completed successful connections. Traffic Relayed ↓ 289543 KB (80.43 KB/s), ↑ 54327 KB (15.09 KB/s).
2026/01/28 18:36:23 In the last 1h0m0s, there were 33 completed successful connections. Traffic Relayed ↓ 232109 KB (64.47 KB/s), ↑ 43652 KB (12.13 KB/s).
2026/01/28 17:36:23 In the last 1h0m0s, there were 27 completed successful connections. Traffic Relayed ↓ 189321 KB (52.59 KB/s), ↑ 35614 KB (9.89 KB/s).
2026/01/28 16:36:23 In the last 1h0m0s, there were 49 completed successful connections. Traffic Relayed ↓ 356712 KB (99.09 KB/s), ↑ 67089 KB (18.64 KB/s).
2026/01/28 15:36:23 In the last 1h0m0s, there were 39 completed successful connections. Traffic Relayed ↓ 275432 KB (76.51 KB/s), ↑ 51743 KB (14.37 KB/s).
2026/01/28 14:36:23 In the last 1h0m0s, there were 45 completed successful connections. Traffic Relayed ↓ 318765 KB (88.55 KB/s), ↑ 59876 KB (16.63 KB/s).
2026/01/28 13:36:23 In the last 1h0m0s, there were 34 completed successful connections. Traffic Relayed ↓ 239654 KB (66.57 KB/s), ↑ 45021 KB (12.51 KB/s).
2026/01/28 12:36:23 In the last 1h0m0s, there were 26 completed successful connections. Traffic Relayed ↓ 182347 KB (50.65 KB/s), ↑ 34298 KB (9.53 KB/s).
2026/01/28 11:36:23 In the last 1h0m0s, there were 51 completed successful connections. Traffic Relayed ↓ 367891 KB (102.19 KB/s), ↑ 69123 KB (19.20 KB/s).`;
			return new Response(mockLogs, {
				headers: { "Content-Type": "text/plain" },
			});
		}

		const filePath = url.pathname === "/" ? "./index.html" : `.${url.pathname}`;
		const file = Bun.file(filePath);

		if (await file.exists()) {
			return new Response(file);
		}

		return new Response("Not Found", { status: 404 });
	},
});

console.log(`Dev server running at http://localhost:${server.port}`);

File: dev-server.go
//go:build dev

package main

import (
	"log"
	"net/http"
	"os"
)

const mockMetrics = `# HELP tor_snowflake_proxy_connection_timeouts_total The total number of client connection attempts that failed after successful rendezvous. Note that failures can occur for reasons outside of the proxy's control, such as the client's NAT and censorship situation.
# TYPE tor_snowflake_proxy_connection_timeouts_total counter
tor_snowflake_proxy_connection_timeouts_total 1649
# HELP tor_snowflake_proxy_connections_total The total number of successful connections handled by the snowflake proxy
# TYPE tor_snowflake_proxy_connections_total counter
tor_snowflake_proxy_connections_total{country=""} 110
tor_snowflake_proxy_connections_total{country="??"} 11
tor_snowflake_proxy_connections_total{country="AE"} 4
tor_snowflake_proxy_connections_total{country="AM"} 1
tor_snowflake_proxy_connections_total{country="AO"} 1
tor_snowflake_proxy_connections_total{country="AR"} 1
tor_snowflake_proxy_connections_total{country="AT"} 3
tor_snowflake_proxy_connections_total{country="AU"} 9
tor_snowflake_proxy_connections_total{country="AZ"} 1
tor_snowflake_proxy_connections_total{country="BD"} 2
tor_snowflake_proxy_connections_total{country="BE"} 1
tor_snowflake_proxy_connections_total{country="BF"} 2
tor_snowflake_proxy_connections_total{country="BG"} 1
tor_snowflake_proxy_connections_total{country="BR"} 1
tor_snowflake_proxy_connections_total{country="BY"} 3
tor_snowflake_proxy_connections_total{country="CA"} 4
tor_snowflake_proxy_connections_total{country="CD"} 1
tor_snowflake_proxy_connections_total{country="CG"} 1
tor_snowflake_proxy_connections_total{country="CH"} 3
tor_snowflake_proxy_connections_total{country="CI"} 4
tor_snowflake_proxy_connections_total{country="CN"} 15
tor_snowflake_proxy_connections_total{country="CO"} 1
tor_snowflake_proxy_connections_total{country="CY"} 1
tor_snowflake_proxy_connections_total{country="DE"} 11
tor_snowflake_proxy_connections_total{country="DK"} 1
tor_snowflake_proxy_connections_total{country="EG"} 7
tor_snowflake_proxy_connections_total{country="EU"} 1
tor_snowflake_proxy_connections_total{country="FI"} 6
tor_snowflake_proxy_connections_total{country="FR"} 10
tor_snowflake_proxy_connections_total{country="GA"} 3
tor_snowflake_proxy_connections_total{country="GB"} 16
tor_snowflake_proxy_connections_total{country="GH"} 3
tor_snowflake_proxy_connections_total{country="GR"} 2
tor_snowflake_proxy_connections_total{country="HK"} 1
tor_snowflake_proxy_connections_total{country="HN"} 1
tor_snowflake_proxy_connections_total{country="ID"} 1
tor_snowflake_proxy_connections_total{country="IL"} 1
tor_snowflake_proxy_connections_total{country="IN"} 6
tor_snowflake_proxy_connections_total{country="IR"} 169
tor_snowflake_proxy_connections_total{country="IT"} 4
tor_snowflake_proxy_connections_total{country="JP"} 2
tor_snowflake_proxy_connections_total{country="KE"} 5
tor_snowflake_proxy_connections_total{country="KH"} 1
tor_snowflake_proxy_connections_total{country="KR"} 1
tor_snowflake_proxy_connections_total{country="LT"} 1
tor_snowflake_proxy_connections_total{country="LV"} 2
tor_snowflake_proxy_connections_total{country="LY"} 1
tor_snowflake_proxy_connections_total{country="MA"} 32
tor_snowflake_proxy_connections_total{country="MG"} 1
tor_snowflake_proxy_connections_total{country="ML"} 1
tor_snowflake_proxy_connections_total{country="MO"} 2
tor_snowflake_proxy_connections_total{country="MU"} 66
tor_snowflake_proxy_connections_total{country="MW"} 5
tor_snowflake_proxy_connections_total{country="MX"} 3
tor_snowflake_proxy_connections_total{country="MY"} 3
tor_snowflake_proxy_connections_total{country="NG"} 9
tor_snowflake_proxy_connections_total{country="NL"} 2
tor_snowflake_proxy_connections_total{country="PE"} 1
tor_snowflake_proxy_connections_total{country="PH"} 2
tor_snowflake_proxy_connections_total{country="PK"} 1
tor_snowflake_proxy_connections_total{country="PL"} 3
tor_snowflake_proxy_connections_total{country="RE"} 1
tor_snowflake_proxy_connections_total{country="RO"} 3
tor_snowflake_proxy_connections_total{country="RS"} 1
tor_snowflake_proxy_connections_total{country="RU"} 327
tor_snowflake_proxy_connections_total{country="RW"} 1
tor_snowflake_proxy_connections_total{country="SA"} 2
tor_snowflake_proxy_connections_total{country="SD"} 8
tor_snowflake_proxy_connections_total{country="SE"} 2
tor_snowflake_proxy_connections_total{country="SG"} 2
tor_snowflake_proxy_connections_total{country="SK"} 1
tor_snowflake_proxy_connections_total{country="SO"} 1
tor_snowflake_proxy_connections_total{country="TH"} 1
tor_snowflake_proxy_connections_total{country="TM"} 5
tor_snowflake_proxy_connections_total{country="TN"} 60
tor_snowflake_proxy_connections_total{country="TR"} 1
tor_snowflake_proxy_connections_total{country="TW"} 1
tor_snowflake_proxy_connections_total{country="UA"} 5
tor_snowflake_proxy_connections_total{country="UG"} 8
tor_snowflake_proxy_connections_total{country="US"} 946
tor_snowflake_proxy_connections_total{country="VN"} 1
tor_snowflake_proxy_connections_total{country="ZA"} 23
tor_snowflake_proxy_connections_total{country="ZM"} 10
# HELP tor_snowflake_proxy_traffic_inbound_bytes_total The total in bound traffic by the snowflake proxy (KB)
# TYPE tor_snowflake_proxy_traffic_inbound_bytes_total counter
tor_snowflake_proxy_traffic_inbound_bytes_total 2.7179552e+07
# HELP tor_snowflake_proxy_traffic_outbound_bytes_total The total out bound traffic by the snowflake proxy (KB)
# TYPE tor_snowflake_proxy_traffic_outbound_bytes_total counter
tor_snowflake_proxy_traffic_outbound_bytes_total 4.204604e+06
# HELP process_network_receive_bytes_total Number of bytes received by the process over the network
# TYPE process_network_receive_bytes_total counter
process_network_receive_bytes_total 4.1371771232e+10
# HELP process_network_transmit_bytes_total Number of bytes sent by the process over the network
# TYPE process_network_transmit_bytes_total counter
process_network_transmit_bytes_total 4.1382248284e+10
# HELP process_resident_memory_bytes Resident memory size in bytes
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 9.5772672e+07
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.76985915868e+09
# HELP go_goroutines Number of goroutines that currently exist
# TYPE go_goroutines gauge
go_goroutines 360
# HELP process_open_fds Number of open file descriptors
# TYPE process_open_fds gauge
process_open_fds 45
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 27268.99
`

const mockLogs = `2026/01/31 12:32:40 In the last 1h0m0s, there were 134 completed successful connections. Traffic Relayed ↓ 183457 KB (50.96 KB/s), ↑ 37294 KB (10.36 KB/s).
2026/01/31 13:32:40 In the last 1h0m0s, there were 114 completed successful connections. Traffic Relayed ↓ 352507 KB (97.92 KB/s), ↑ 101225 KB (28.12 KB/s).
2026/01/31 14:32:40 In the last 1h0m0s, there were 106 completed successful connections. Traffic Relayed ↓ 857188 KB (238.11 KB/s), ↑ 119983 KB (33.33 KB/s).
2026/01/31 15:32:40 In the last 1h0m0s, there were 93 completed successful connections. Traffic Relayed ↓ 561858 KB (156.07 KB/s), ↑ 119705 KB (33.25 KB/s).
2026/01/31 16:32:40 In the last 1h0m0s, there were 86 completed successful connections. Traffic Relayed ↓ 733619 KB (203.78 KB/s), ↑ 127004 KB (35.28 KB/s).
2026/01/31 17:32:40 In the last 1h0m0s, there were 128 completed successful connections. Traffic Relayed ↓ 597328 KB (165.92 KB/s), ↑ 121182 KB (33.66 KB/s).
2026/01/31 18:32:40 In the last 1h0m0s, there were 121 completed successful connections. Traffic Relayed ↓ 348280 KB (96.74 KB/s), ↑ 59854 KB (16.63 KB/s).
2026/01/31 19:32:40 In the last 1h0m0s, there were 120 completed successful connections. Traffic Relayed ↓ 372636 KB (103.51 KB/s), ↑ 50047 KB (13.90 KB/s).
2026/01/31 20:32:40 In the last 1h0m0s, there were 113 completed successful connections. Traffic Relayed ↓ 474974 KB (131.94 KB/s), ↑ 93429 KB (25.95 KB/s).
2026/01/31 21:32:40 In the last 1h0m0s, there were 48 completed successful connections. Traffic Relayed ↓ 729185 KB (202.55 KB/s), ↑ 92820 KB (25.78 KB/s).
2026/01/31 22:32:40 In the last 1h0m0s, there were 16 completed successful connections. Traffic Relayed ↓ 1187745 KB (329.93 KB/s), ↑ 113682 KB (31.58 KB/s).
2026/01/31 23:32:40 In the last 1h0m0s, there were 7 completed successful connections. Traffic Relayed ↓ 715292 KB (198.69 KB/s), ↑ 59830 KB (16.62 KB/s).
2026/02/01 00:32:40 In the last 1h0m0s, there were 8 completed successful connections. Traffic Relayed ↓ 517076 KB (143.63 KB/s), ↑ 32023 KB (8.90 KB/s).
2026/02/01 01:32:40 In the last 1h0m0s, there were 5 completed successful connections. Traffic Relayed ↓ 536223 KB (148.95 KB/s), ↑ 35168 KB (9.77 KB/s).
2026/02/01 02:32:40 In the last 1h0m0s, there were 9 completed successful connections. Traffic Relayed ↓ 635409 KB (176.50 KB/s), ↑ 36176 KB (10.05 KB/s).
2026/02/01 03:32:40 In the last 1h0m0s, there were 12 completed successful connections. Traffic Relayed ↓ 602633 KB (167.40 KB/s), ↑ 36340 KB (10.09 KB/s).
2026/02/01 04:32:40 In the last 1h0m0s, there were 22 completed successful connections. Traffic Relayed ↓ 588310 KB (163.42 KB/s), ↑ 42610 KB (11.84 KB/s).
2026/02/01 05:32:40 In the last 1h0m0s, there were 19 completed successful connections. Traffic Relayed ↓ 1207049 KB (335.29 KB/s), ↑ 87281 KB (24.24 KB/s).
2026/02/01 06:32:40 In the last 1h0m0s, there were 19 completed successful connections. Traffic Relayed ↓ 862143 KB (239.48 KB/s), ↑ 84283 KB (23.41 KB/s).
2026/02/01 07:32:40 In the last 1h0m0s, there were 25 completed successful connections. Traffic Relayed ↓ 994962 KB (276.38 KB/s), ↑ 87620 KB (24.34 KB/s).
2026/02/01 08:32:40 In the last 1h0m0s, there were 27 completed successful connections. Traffic Relayed ↓ 821964 KB (228.32 KB/s), ↑ 105267 KB (29.24 KB/s).
2026/02/01 09:32:40 In the last 1h0m0s, there were 28 completed successful connections. Traffic Relayed ↓ 811943 KB (225.54 KB/s), ↑ 91503 KB (25.42 KB/s).
2026/02/01 10:32:40 In the last 1h0m0s, there were 36 completed successful connections. Traffic Relayed ↓ 1523789 KB (423.27 KB/s), ↑ 108411 KB (30.11 KB/s).
2026/02/01 11:32:40 In the last 1h0m0s, there were 19 completed successful connections. Traffic Relayed ↓ 1858789 KB (516.33 KB/s), ↑ 129278 KB (35.91 KB/s).`

func main() {
	fs := http.FileServer(http.Dir("."))
	http.Handle("/", fs)

	http.HandleFunc("/api/metrics", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/plain")
		w.Write([]byte(mockMetrics))
	})

	http.HandleFunc("/api/logs", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/plain")
		w.Write([]byte(mockLogs))
	})

	http.HandleFunc("/api/nat", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/plain")
		w.Write([]byte("restricted"))
	})

	port := os.Getenv("PORT")
	if port == "" {
		port = "3000"
	}

	log.Printf("Dev server running at http://localhost:%s\n", port)
	log.Fatal(http.ListenAndServe(":"+port, nil))
}
